File: 0054-spiral-matrix/0054-spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        int m = matrix.length;
        int n = matrix[0].length;

        int[] right = new int[]{0,1}; 
        int[] left = new int[]{0,-1}; 
        int[] up = new int[]{-1,0}; 
        int[] down = new int[]{1,0}; 

        int[][] directions = new int[][]{right, down, left, up};
        int pointerDir = 0;

        List<Integer> result = new ArrayList<>();

        int total = m * n;
        int count = 0;
        int visited = -101;
        int i = 0, j = 0;
        while (count < total){
            result.add(matrix[i][j]);
            matrix[i][j] = visited;
            count++;
            
            // Update
            int nextRow = i + directions[pointerDir][0];
            int nextCol = j + directions[pointerDir][1];

            if (nextRow >= m || nextCol >= n || nextCol < 0 ||
                matrix[nextRow][nextCol] == visited) 
            {
                pointerDir = (pointerDir + 1) % directions.length;
            }
            i += directions[pointerDir][0];
            j += directions[pointerDir][1];
        }

        return result;
    }
}

File: 0054-spiral-matrix/0054-spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        dirs : List[Tuple[int,int]] = [(0,1), (1,0), (0,-1), (-1,0)]
        dir : int = 0

        m : int = len(matrix)
        n : int = len(matrix[0])

        output : List[int] = []
        i : int = 0 
        j : int = 0
        visited = set()
        while i < m and j < n and len(visited) < m*n:
            visited.add((i,j))
            output.append(matrix[i][j])
            new_i : int = i + dirs[dir][0]
            new_j : int = j + dirs[dir][1]
            if (new_i >= m or new_j >= n or new_i < 0 or new_j < 0) or (new_i,new_j) in visited:
                dir = (dir + 1) % 4
                new_i = i + dirs[dir][0]
                new_j = j + dirs[dir][1]
            
            i = new_i
            j = new_j
        return output
